### PR TITLE
Allow nested fields in the composite aggregation

### LINKED ISSUE #28611: composite aggregation as child aggregation
The 6.x new composite aggregation provides a way to scroll and page on terms bucket, and it's a good thing.
However, it's not (for now?) possible to set a composite aggregation as a child of another bucket aggregation.
So it's impossible, for example, to composite-aggregate on a nested or a child entity. This request is not allowed :

```
{
        "metadatas" : {
            "nested" : {
                "path" : "metaDatas"
            },
            "aggs" : {
                "msisdn" : {
                  "filter" : { "term": { "metaDatas.name": "msisdn" } },
                  "aggs": {
                    "msisdns": {
                       "composite" : {
                          "size":5,
                          "sources" : [{ "value": { "terms" : { "field": "metaDatas.value.raw" } } }]
                       }
                    }
                  }
                }
            }
        }
    }
```
I don't understand this limitation. 
Furthermore, an ES 6.2 patched without the check seems to work as expected.
 ```
@Override
    protected AggregatorFactory<?> doBuild(SearchContext context, AggregatorFactory<?> parent,
                                           AggregatorFactories.Builder subfactoriesBuilder) throws IOException {
        /*if (parent != null) {
            throw new IllegalArgumentException("[composite] aggregation cannot be used with a parent aggregation");
        }*/
```
any insight ?

### PR DESCRIPTION
**Pull request by jimczi**
This changes adds the support to handle `nested` fields in the `composite`
aggregation. A `nested` aggregation can be used as parent of a `composite`
aggregation in order to target `nested` fields in the `sources`.

Closes #28611

### CODE REVIEW

File: server/src/main/java/org/elasticsearch/search/aggregations/bucket/composite/CompositeAggregationBuilder.java [hunk 1]
```diff
@@ -151,11 +152,28 @@ public int size() {
...
+        }
+    }
+
     @Override
     protected AggregatorFactory<?> doBuild(SearchContext context, AggregatorFactory<?> parent,
                                            AggregatorFactories.Builder subfactoriesBuilder) throws IOException {
-        if (parent != null) {
-            throw new IllegalArgumentException("[composite] aggregation cannot be used with a parent aggregation");
+        AggregatorFactory<?> invalid = checkParentIsNullOrNested(parent);
+        if (invalid != null) {
```
Comment (polyfractal) [inline server/src/main/java/org/elasticsearch/search/aggregations/bucket/composite/CompositeAggregationBuilder.java]: Out of curiosity, what scenario might trigger this now that nested is supported?

**Review by polyfractal — APPROVED** [review]
LGTM, sorry for the delay!

File: server/src/main/java/org/elasticsearch/search/aggregations/bucket/composite/CompositeAggregationBuilder.java (hunk 1 above)
Comment (jimczi) [reply server/src/main/java/org/elasticsearch/search/aggregations/bucket/composite/CompositeAggregationBuilder.java]: If you add a parent other than a `nested` aggregator. So for instance a `terms` aggregation cannot use a `composite` aggregation as a sub-aggregation.

**Comment by krishan-21723** [comment]
why can't it support all kinds of aggregators as parent ?

**Comment by edoquiros** [comment]
Starting on which version is this fix available?

**Pull request merged**